enhancements/machine-config/pin-and-pre-load-images.md

```diff
@@ -0,0 +1,349 @@
+    matchLabels:
+      node-role.kubernetes.io/control-plane: ""
+  pinnedImages:
+  ...
+
+---
+
+# For worker nodes:
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: PinnedImageSet
+metadata:
+  name: my-control-plane-pinned-images
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker: ""
+  pinnedImages:
+  ...
+```
+
+This is specially convenient for pinning images for upgrades: there are many
+images that are needed only by the control plane nodes and there is no need to
+have them consuming disk space in worker nodes.
+
+When no node selector is specified the images will be pinned in all the nodes
+of the cluster.
+
+When a `PinnedImageSet` custom resource is added, modified or deleted the
+machine config operator will create, modify or delete the configuration file,
+reload the CRI-O configuration (with the equivalent of `systemctl reload crio`)
+and then it will use the CRI-O gRPC API to run the equivalent of `crictl pull`
+for each of the images.
+
+Note that currently CRI-O doesn't reset pinned images on reload, support for
+that will need to be added.
+
+Note that this will happen in all the nodes of the cluster that match the node
+selector.
+
+When all the images have been succesfully pinned and pulled in all the matching
+nodes the `Ready` condition will be set to `True`:
+
+```yaml
+status:
+  conditions:
+  - type: Ready
+    status: "True"
+  - type: Failed
+    status: "False"
+```
+
+If something fails the `Failed` condition will be set to `True`, and the
+details of the error will be in the message. For example if `node12` fails to
+pull an image:
+
+```yaml
+status:
+  pinnedImages:
+  - quay.io/openshift-release-dev/ocp-release@sha256:...
+  - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+  conditions:
+  - type: Ready
+    status: "False"
+  - type: Failed
+    status: "True"
+    message: Node 'node12' failed to pull image `quay.io/...` because ...
+```
+
+This logic to handle the pinned image sets and generate the configuration files
+will be part of a new sub-controller in MCC.
+
+### Risks and Mitigations
+
+Pre-loading disk images can consume a large amount of disk space. For example,
+pre-loading all the images required for an upgrade can consume more 32 GiB.
+This is a risk because disk exhaustion can affect other workloads and the
+control plane. There is already a mechanism to mitigate that: the kubelet
+garbage collection. To ensure disk space issues are reported and that garbage
+collection doesn't interfere we will introduced new mechanisms:
+
+1. If disk exhaustion happens while trying to pre-load images the issues will
+be reported explicitly via the status of the `PinnedImageSet` status. Note
+typically these issues are reported as failures to pull images in the status of
+pods, but in this case there are no pods pulling the images. CRI-O will still
+report these issues in the log (if there is space for that), like for any other
+image pull.
```

Comment:
Do we then do anything with the PIS crio config? Do we just keep trying to pull pinned images hoping that garbage collection frees sufficient space to complete the operation?

If we delete it then that doesn't really free the space until cri-o gets restarted per earlier notes about resetting. This worries me that we could wedge nodes, should we ensure that before we pull images we compute the space which the garbage collection threshold would trigger at and abort if the content we'd pull exceeds the minimum free space available and not pull if GC is in progress?

32GiB to pull, 320GiB disk with 10% GC threshold wouldn't allow us to pull, but any threshold higher than that would because we can be sure if GC is not running we've got enough space to complete our operation.

Reply:
We intend to add the upgrade orchestration in a separate EP, based on #1432. I would like to do after this and #1483 are approved.

My understanding is that "reset" in this context means that when CRI-O receives the reload signal it needs to forget the previous set of pinned images and rebuild it from the configuration files. According to @haircommander it doesn't do that today.

Note that our suggestion is that when some `PinnedImageSet`  changes first CRI-O is reconfigured and reloaded, and only then images start to be pulled. Imagine, for example, that you have images A and B pinned and then you remove B and add C. What would happen first is that CRI-O is reconfigured so that only A and C are pinned. That is unlikely to fail, as it only involves generating configuration files and sending signals. Only after that would we start to pull the images. That is more likely to fail, for example because C cannot be pulled. But even if that fails Kubelet can still garbage collect B.

I'd say that if disk is exhausted while pulling the images the behavior should be similar to what happens when a regular pod fails to pull an image. We retry that, so we should retry this as well.

Computing the disk space needed in advance would be helpful indeed. But implementing it will be challenging. It means contacting the registry to figure out the size of the layers, but also figuring out how much space those layers will take when extracted, and that requires intimate knowledge of the underlying workings of the container library. Would it be acceptable to use an heuristic? For example, can we contact the registry to find the sizes of the layers and then assume that the space consumed will be twice that? I say twice because that is roughly the case of the current OCP payloads. @sdodson let me know of you think that is acceptable.

Reply:
Yeah, starting with 2x the size of the layers seems like a good start to me.